[PlayStation] Start RemoteInspectorServer

haruhisa-shin opened the pull request:
#### 650c693e1678cb6f0fd0a31353f959ab204473eb
<pre>
[PlayStation] Start RemoteInspectorServer
<a href="https://bugs.webkit.org/show_bug.cgi?id=256576">https://bugs.webkit.org/show_bug.cgi?id=256576</a>

Reviewed by Devin Rousso and Fujii Hironori.

The MiniBrowser starts RemoteInspectorServer.
Also, fixed build error when ENABLE_REMOTE_INSPECTOR is OFF.

* Source/JavaScriptCore/API/JSRemoteInspectorServer.cpp:
(JSRemoteInspectorServerStart):
* Tools/MiniBrowser/playstation/main.cpp:
(initialize):

Canonical link: <a href="https://commits.webkit.org/264404@main">https://commits.webkit.org/264404@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d23d35a506a5120909ace6ec646b653415210c4a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7398 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7656 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7831 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9027 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7603 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7408 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9513 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7581 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10488 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7527 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8588 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6845 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9136 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5902 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6729 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14455 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6293 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7200 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10097 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6981 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7332 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5995 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7543 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6683 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1740 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1783 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10890 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7748 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7068 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1880 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

haruhisa-shin commented on Source/WebKit/UIProcess/playstation/WebProcessPoolPlayStation.cpp:
```diff
@@ -32,11 +32,47 @@
+    static std::once_flag flag;
+    std::call_once(flag, [] {
+        char buffer[64];
+        if (!getenv_np("WebInspectorServerPort", buffer, sizeof(buffer)))
```
As a side note, the environment variable  `WebInspectorServerPort` is already defined and used in MiniBrowser:
https://github.com/WebKit/WebKit/blob/main/Tools/MiniBrowser/playstation/main.cpp#LL53C16-L53C38

dcrousso replied:
NIT: can we make that into a shared `static constexpr` or something so that we can guarantee it's shared by both?

donny-dont reviewed (commented):
I don't see anything wrong with the code in question I'm just wondering if there's a better way to approach this rather than an environment variable.

dcrousso commented on Source/WebKit/UIProcess/playstation/WebProcessPoolPlayStation.cpp:
```diff
@@ -32,11 +32,47 @@
+    if (!portNumber)
+        return;
+
+    Inspector::RemoteInspectorServer::singleton().start(nullptr, portNumber.value());
```
we don't want/need an `address`?  if we do want/need it, should we add it as a `WebInspectorAddress` envvar?

haruhisa-shin replied:
Yes, we currently use only port and not address.

If we were to use both in the future, we would prefer to use the "ip:port" format and split them in the code, similar to wincairo.

donny-dont commented on Source/WebKit/UIProcess/API/C/WKRemoteInspectorServer.cpp:
```diff
@@ -0,0 +1,40 @@
+
+    return server.getPort().value_or(0);
+}
+#endif // ENABLE(REMOTE_INSPECTOR)
```
An issue is that this will have a linker error if `!ENABLE(REMOTE_INSEPCTOR)`.
```suggestion
#if ENABLE(REMOTE_INSPECTOR)
#include <JavaScriptCore/RemoteInspectorServer.h>
#endif

uint16_t WKRemoteInspectorServerStart(const char* address, uint16_t port)
{
#if ENABLE(REMOTE_INSPECTOR)
    auto& server = Inspector::RemoteInspectorServer::singleton();
    if (!server.start(address, port))
        return 0;

    return server.getPort().value_or(0);
#else
    return 0;
#endif
}
```

donny-dont commented on Tools/MiniBrowser/playstation/main.cpp:
```diff
@@ -35,6 +35,10 @@
 
+#if defined (ENABLE_REMOTE_INSPECTOR) && ENABLE_REMOTE_INSPECTOR
+#include <WebKit/WKRemoteInspectorServer.h>
+#endif
```
Shouldn't be conditional.

donny-dont commented on Tools/MiniBrowser/playstation/main.cpp:
```diff
@@ -80,6 +83,10 @@ static void initialize()
+
+#if defined (ENABLE_REMOTE_INSPECTOR) && ENABLE_REMOTE_INSPECTOR
+    WKRemoteInspectorServerStart(nullptr, 868);
+#endif
```
Shouldn't be conditional.

donny-dont reviewed (commented):
API users don't know if `ENABLE_REMOTE_INSPECTOR` is on so we need to have an implementation there. Not sure if we want to have a return value that says that its not available other than returning `0`.

Also not sure if it'd be good form to have a sentinel value of `InsepctorNotStarted` or something like that instead of 0. @fujii any thoughts?

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKRemoteInspectorServer.h:
```diff
@@ -0,0 +1,38 @@
+extern "C" {
+#endif
+
+WK_EXPORT uint16_t WKRemoteInspectorServerStart(const char* address, uint16_t port);
```
I would expect a corresponding "Stop" call.
What does "address" expect?  I only see nullptr being used with it.

haruhisa-shin replied:
About Stop:
We have never had to stop a server.
Therefore, the socket implementation is not designed to be able to stop.
This may be a major change because it makes no sense to be able to restart the server if it is stopped.
I think it would be better to implement `start` first and make the implementation of `stop` a separate task.

haruhisa-shin replied:
About address:
As you said, the address is omittable because it is usually nullptr(=INADDR_ANY) or "0.0.0.0" for playstation port.
However, GTK and wincairo do not ignore the address specified in the environment variable and assign it to the socket, so I did the same.

fujii commented on Source/WebKit/UIProcess/API/C/WKRemoteInspectorServer.cpp:
```diff
@@ -0,0 +1,40 @@
+#if ENABLE(REMOTE_INSPECTOR)
+#include <JavaScriptCore/RemoteInspectorServer.h>
+
+uint16_t WKRemoteInspectorServerStart(const char* address, uint16_t port)
```
This code looks exactly same with `JSRemoteInspectorServerStart`. What will happend if you call `JSRemoteInspectorServerStart` instead of `WKRemoteInspectorServerStart`?
https://github.com/WebKit/WebKit/blob/23bae944192123a739e1519b96081ec235e73800/Source/JavaScriptCore/API/JSRemoteInspectorServer.cpp#L32-L39

BTW, I think we should add a new API `JSRemoteInspectorServerStop` to stop the server.

haruhisa-shin replied:
>  What will happend if you call JSRemoteInspectorServerStart instead of WKRemoteInspectorServerStart?

The results will be exactly the same.

fujii replied:
If so, do we need the new API?

haruhisa-shin replied:
That reminds me.
For some reason, I had assumed that the JS_EXPORT functions should not be called from the MiniBrowser.
Just to confirm, what you are saying is that we can call `JSRemoteInspectorServerStart` directly here, is that correct?

fujii replied:
> That reminds me. For some reason, I had assumed that the JS_EXPORT functions should not be called from the MiniBrowser. 

Why?

> Just to confirm, what you are saying is that we can call `JSRemoteInspectorServerStart` directly here, is that correct?

No.
I'd like to know how do you think about an idea using the existing API instead of adding same API.

haruhisa-shin replied:
> Why?

As I said above, "for some reason". @fujii -san's comment helped me realize that was a wrong thought.

> I'd like to know how do you think about an idea using the existing API instead of adding same API.

Since RemoteInspectorServer is part of JSC, I think it is appropriate to start it via `JSRemoteInspectorServerStart()`.
If it's not a problem, I'm willing to change it that way, but would appreciate your opinion.

fujii reviewed (approved):
LGTM

haruhisa-shin committed:
[PlayStation] Start RemoteInspectorServer
https://bugs.webkit.org/show_bug.cgi?id=256576

Reviewed by Devin Rousso and Fujii Hironori.

The MiniBrowser starts RemoteInspectorServer.
Also, fixed build error when ENABLE_REMOTE_INSPECTOR is OFF.

* Source/JavaScriptCore/API/JSRemoteInspectorServer.cpp:
(JSRemoteInspectorServerStart):
* Tools/MiniBrowser/playstation/main.cpp:
(initialize):

Canonical link: https://commits.webkit.org/264404@main
